oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServletOAuth2AuthorizedClientExchangeFilterFunctionTests.java

```diff
@@ -635,6 +643,101 @@ public void filterWhenRequestAttributesNotSetAndHooksInitHooksResetThenDefaultsN
 		assertThat(getBody(request)).isEmpty();
 	}
 
+	// gh-7228
+	@Test
+	public void publisherWhenHooksInitAndSubscribedOutsideOfWebContextThenShouldNotThrowNpe() throws Exception {
```

Comment:
If I comment out the following in `createRequestContextSubscriber()`

```
//		if (authentication == null && request == null && response == null) {
//			return delegate;
//		}
```

....the test still passes. I'm not sure this test makes sense?

Reply:
because this test does not checks this part (publisherWhenHooksInitAndSubscribedOutsideOfWebContextAndOutsideOfSecurityContextAndWithParentContextThenShouldNotReplaceParentContextAndShouldNotAddRequestContext() tests it)   
The check you commented is intended to skip reactor context population with empty RequestContextDataHolder 
 values as it is not useful data (especially outside of web context). This is just optimization ( here my motivation https://github.com/spring-projects/spring-security/pull/7235#issuecomment-520127370 )

Reply:
NPE is fixed by introducing RequestContextDataHolder which could contain null values,   
It also reduces number of occupied keys in reactor context